src/storage/kv/transaction.rs

```diff
@@ -0,0 +1,1163 @@
+
+        {
+            let txn3 = store.begin().unwrap();
+            let val1 = txn3.get(&key1).unwrap();
+            assert_eq!(val1, value1.as_ref());
+            let val2 = txn3.get(&key2).unwrap();
+            assert_eq!(val2, value2.as_ref());
+        }
+    }
+
+    // G1b: Intermediate Reads (dirty reads)
+    #[test]
+    fn test_snapshot_isolation_g1b() {
+        let store = create_store();
+
+        let key1 = Bytes::from("k1");
+        let key2 = Bytes::from("k2");
+        let value1 = Bytes::from("v1");
+        let value3 = Bytes::from("v3");
+        let value4 = Bytes::from("v4");
+
+        {
+            let mut txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key1).is_ok());
+            assert!(txn1.get(&key2).is_ok());
+
+            txn1.set(&key1, &value3).unwrap();
+
+            let range = VectorKey::from_str("k1")..=VectorKey::from_str("k3");
+            let res = txn2.scan(range.clone()).unwrap();
+            assert_eq!(res.len(), 2);
+            assert_eq!(res[0].0, value1);
+
+            txn1.set(&key1, &value4).unwrap();
+            txn1.commit().unwrap();
+
+            let res = txn2.scan(range).unwrap();
+            assert_eq!(res.len(), 2);
+            assert_eq!(res[0].0, value1);
+
+            txn2.commit().unwrap();
+        }
+    }
+
+    // G1c: Circular Information Flow (dirty reads)
+    #[test]
+    fn test_snapshot_isolation_g1c() {
+        let store = create_store();
+
+        let key1 = Bytes::from("k1");
+        let key2 = Bytes::from("k2");
+        let value1 = Bytes::from("v1");
+        let value2 = Bytes::from("v2");
+        let value3 = Bytes::from("v3");
+        let value4 = Bytes::from("v4");
+
+        {
+            let mut txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key1).is_ok());
+            assert!(txn2.get(&key2).is_ok());
+
+            txn1.set(&key1, &value3).unwrap();
+            txn2.set(&key2, &value4).unwrap();
+
+            assert_eq!(txn1.get(&key2).unwrap(), value2.as_ref());
+            assert_eq!(txn2.get(&key1).unwrap(), value1.as_ref());
+
+            txn1.commit().unwrap();
+            assert!(match txn2.commit() {
+                Err(err) => {
+                    if let Error::TxnReadConflict = err {
+                        true
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            });
+        }
+    }
+
+    // PMP: Predicate-Many-Preceders
+    #[test]
+    fn test_snapshot_isolation_pmp() {
+        let store = create_store();
+
+        let key2 = Bytes::from("k2");
+        let key3 = Bytes::from("k3");
+        let value3 = Bytes::from("v3");
+
+        {
+            let txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key3).is_err());
```

Comment:
Shouldn't this and the same op in L947 be a scan against `v1..=v3`?
It seems to me that currently two txn1's read sets and the txn2's write set don't overlap so there might be no chance of an anomaly even if it wasn't based on MVCC?

Reply:
Also, what kind of error are we currently expecting here? (A missing key error, by any chance?

For a PMP test, I was expecting we would check that two txn1 scans before and after the key insertion made by the txn2 give the same result sets.

But again, I might be misunderstanding something badly! Maybe you could give me a brief explanation of what you are trying to check with this test, so that I can figure it out myself.

Reply:
Ah, my bad. The error is KeyNotFound. 

1) I will add the explicit error assertion to make it more clear.
2) With regards to scan, I did not use it because there is no predicate search as such yet in the KV store. The tests in hermitage usually are doing a select based on value, so I wasn't sure of adding a scan to read all the keys. 

This is the [reference](https://github.com/ept/hermitage/blob/master/foundationdb.md)

If you think scans will make it more explicit, let me know. I wanted to search for a specific key (searching for key3) to make it more readable.

Reply:
Thanks! Re point 2, I think you'd rather test for imp rather than pmp then.

It looks like to me that pmp is a specialization(?) of imp; if you take the predicate used here as "match all (a predicate returns true for anything)", used `scan` instead of `get` for testing it, it would look much more like what is described in `pmp` test described in https://github.com/ept/hermitage/blob/master/foundationdb.md and https://sitano.github.io/theory/databases/2019/07/30/tx-isolation-anomalies/#imppmp-predicate-many-preceders

Currently, this "test_snapshot_isolation_pmp" looks like testing something like the below if written in a hermitage-like way:

```
    tell 0 "begin"
    tell 1 "begin"
    tell 0 "select * from test where key = 'key3'" # Returns nothing
    tell 1 "insert into test (key, value) values('key2', 'value3')" # You write to another key
    tell 1 "commit"
    tell 0 "select * from test where key = 'key3'" # Still returns nothing
    tell 0 "commit"
```

The readset and writeset doesn't intersect so this would pass even when the implementation doesn't actually prevent pmp and even imp.

If we are to test for imp for now as we don't have predicates in surrealkv, a hermitage-style test would look like:


```
    tell 0 "begin"
    tell 1 "begin"
    tell 0 "select * from test" # Returns nothing
    tell 1 "insert into test (key, value) values('key2', 'value3')" # You write to another key
    tell 1 "commit"
    tell 0 "select * from test" # Still returns nothing
    tell 0 "commit"
```

Notice that I just omitted `where key = 'key3'`. Doing so would correct using `scan` in this test? 🤔

Reply:
Or.. it could look a bit more like PMP if txn1 and txn2 reads and writes from/to the same key, assuming the predicate is a equality check against the key of the kv pairs.

Maybe use `scan` in imp test and change the txn1 and txn2 to readwrite the same key for pmp test?

Reply:
Thanks for this @mumoshu, I have made the test as you have suggested. Do you think renaming it to IMP would be better?

Reply:
I'll keep a PR open for this if you think renaming this to IMP is right Yusuke!